crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+//! Implements user space page fault handling with the `userfaultfd` ("uffd") system call on Linux.
+//!
+//! Handling page faults for memory accesses in regions relating to WebAssembly instances
+//! enables the implementation of protecting guard pages in user space rather than kernel space.
+//!
+//! This reduces the number of system calls and kernel locks needed to provide correct
+//! WebAssembly memory semantics.
+//!
+//! Additionally, linear memories can be lazy-initialized upon access.
+//!
+//! This feature requires a Linux kernel 4.11 or newer to use.
+
+use super::{InstancePool, StackPool};
+use crate::{instance::Instance, Mmap};
+use std::convert::TryInto;
+use std::ptr;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use std::thread;
+use userfaultfd::{Event, FeatureFlags, IoctlFlags, Uffd, UffdBuilder};
+use wasmtime_environ::{entity::EntityRef, wasm::DefinedMemoryIndex, MemoryInitialization};
+
+const WASM_PAGE_SIZE: usize = wasmtime_environ::WASM_PAGE_SIZE as usize;
+
+pub unsafe fn make_accessible(_addr: *mut u8, _len: usize) -> bool {
+    // A no-op when userfaultfd is used
+    true
+}
+
+pub unsafe fn reset_guard_page(addr: *mut u8, len: usize) -> bool {
+    // Guard pages are READ_WRITE with uffd until faulted
+    region::protect(addr, len, region::Protection::READ_WRITE).is_ok()
```

Comment:
I'll admit I unfortunately have no idea what the purpose of this function nor why this (presumably `mprotect`?) call does what it needs to do. Could you explain a bit more what's going on with this side table of guard pages?

Reply:
This resets the previously faulted guard pages to the expected read/write permission; the entire range of pages monitored by the user fault handler is read/write so a fault can occur anywhere in that range and the kernel will pass it to us.

This call to `mprotect` only ever occurs when a guard page has faulted; if a memory access is never out of bounds, then the memory managed by the pooling allocator will never have its protection level changed while still providing the correct semantics for Wasm.

Reply:
So to make sure I understand

* Guard pages *never* have physical memory backing them (e.g. `NORESERVE`)
* If permission is read/write, faults get routed to our uffd
* If permission is `NONE`, faults get routed to the signal handler

So this reset needs to happen so future faults go to uffd?

Reply:
Your summary is correct.

If we limit the recording of resettable guard page to only pages that could potentially be accessible in the future, the reset is basically telling the kernel memory manager that we want to start receiving the faults again for that page.  Upon the next fault we then get to decide to back the page or treat it once against as a guard page.